diesel/src/query_builder/select_by_statement/dsl_impls.rs

```diff
@@ -0,0 +1,362 @@
+// use super::BoxedSelectByStatement;
+use crate::associations::HasTable;
+// use crate::backend::Backend;
+// use crate::expression::nullable::Nullable;
+use crate::expression::*;
+use crate::insertable::Insertable;
+use crate::query_builder::insert_statement::InsertFromSelect;
+use crate::query_builder::{Query, SelectQuery, SelectByStatement};
+// use crate::query_dsl::boxed_dsl::BoxedDsl;
+use crate::query_dsl::methods::*;
+use crate::query_dsl::*;
+// use crate::query_source::joins::{Join, JoinOn, JoinTo};
+use crate::query_source::joins::JoinTo;
+// use crate::query_source::QuerySource;
+use crate::sql_types::Bool;
+
+impl<S, STMT, Rhs, Kind, On> InternalJoinDsl<Rhs, Kind, On>
+    for SelectByStatement<S, STMT>
+where
+    STMT: InternalJoinDsl<Rhs, Kind, On>,
+{
+    type Output = STMT::Output;
+
+    fn join(self, rhs: Rhs, kind: Kind, on: On) -> Self::Output {
+        self.inner.join(rhs, kind, on)
+    }
+}
+
+impl<S, STMT, Selection> SelectDsl<Selection>
+    for SelectByStatement<S, STMT>
+where
+    Selection: Expression,
+    STMT: SelectDsl<Selection>,
+{
+    type Output = STMT::Output;
+
+    fn select(self, selection: Selection) -> Self::Output {
+        self.inner.select(selection)
+    }
+}
+
+impl<ST, S, STMT> DistinctDsl for SelectByStatement<S, STMT>
+where
+    Self: SelectQuery<SqlType = ST>,
```

Comment:
1. Should we check `SelectQuery` in every dsl?
2. Is there any better solution to check inner `SelectClause` not changed (internally).

Reply:
> 1. Should we check SelectQuery in every dsl?

I don't think that's required.

> 2. Is there any better solution to check inner SelectClause not changed (internally).

I do not think that adding a distinct "clause" to the query will change the fields returned by the query, so I'm not sure we even need to check that.

Reply:
I use `SelectByQuery` to ensure inner Columns invariant. We don't force inner STMT to be SelectStatement